packages/gatsby-dev-cli/src/verdaccio.js

```diff
@@ -0,0 +1,137 @@
+const startVerdaccio = require(`verdaccio`).default
+const path = require(`path`)
+const npmLogin = require(`npm-cli-login`)
+const fs = require(`fs-extra`)
+const { promisifiedSpawn } = require(`./utils`)
+
+let VerdaccioInitPromise = null
+
+const verdaccioConfig = {
+  storage: path.join(__dirname, `..`, `verdaccio`, `storage`),
+  port: 4873, // default
+  web: {
+    enable: true,
+    title: `gatsby-dev`,
+  },
+  auth: {
+    htpasswd: {
+      file: `./.htpasswd`,
+      max_users: 1000,
+    },
+  },
+  packages: {
+    "**": {
+      access: `$all`,
+      publish: `gatsby-dev`, // we create dummy gatsby-dev username
+      proxy: `npmjs`,
+    },
+  },
+  uplinks: {
+    npmjs: {
+      url: `https://registry.npmjs.org/`,
+    },
+  },
+  self_path: path.join(__dirname, `..`, `verdaccio`),
+}
+
+const registryUrl = `http://localhost:${verdaccioConfig.port}`
+
+exports.registryUrl = registryUrl
+
+const startServer = () => {
+  if (VerdaccioInitPromise) {
+    return VerdaccioInitPromise
+  }
+
+  console.log(`Starting local verdaccio server`)
+
+  // clear storage
+  fs.removeSync(verdaccioConfig.storage)
+
+  VerdaccioInitPromise = new Promise(resolve => {
+    startVerdaccio(
+      verdaccioConfig,
+      verdaccioConfig.port,
+      verdaccioConfig.storage,
+      `1.0.0`,
+      `gatsby-dev`,
+      (webServer, addr, pkgName, pkgVersion) => {
+        webServer.listen(addr.port || addr.path, addr.host, () => {
+          console.log(`Started local verdaccio server`)
+
+          // Login with dummy gatsby-dev user - see  ./verdaccio/.htpasswd .
+          // This is needed to be able to actually publish packages
+          // to local registry.
+          console.log(`Logging in with dummy user`)
+
+          npmLogin(
+            `gatsby-dev`,
+            `gatsby-dev`,
+            `gatsby-dev@gatsbyjs.org`,
+            registryUrl
+          )
+
+          // This is silly - `npm-cli-login` doesn't return promise or callback
+          // so let's just wait reasaonable amount before resolving
+          // There is open pull request to add returning promise:
+          // https://github.com/postmanlabs/npm-cli-login/pull/19
+          setTimeout(() => {
+            console.log(`Hopefully logged in already`)
+            resolve()
+          }, 1500)
+        })
+      }
+    )
+  })
+
+  return VerdaccioInitPromise
+}
+
+exports.publishPackageLocally = async ({
+  pathToPackage,
+  packageName,
+  version,
+}) => {
+  // <ake sure verdaccio is running
+  await startServer()
+
+  // npm unpublish
+  // because version doesn't change when developing locally verdaccio wouldn't
+  // let republish package with same version - so as quick workaround we
+  // unpublish package.
+  // Ideally we could check if we need to unpublish: i.e. compare deps of
```

Comment:
If you force publish `npm publish --force` the version is replaced. Perhaps that helps.

Reply:
ohhh, cool, thanks for the tip! But there is another issue that will render this a bit irrelevant - let me write up follow up comment to document what I mean